File: impl/ts/src/client/rekey.ts
/**
 * Background rekey driver. Mirrors `impl/go/internal/client/rekey.go`:
 * fires once at 80% of the session TTL, then exits.
 *
 * @module
 */

import { rekeyClient } from "@sempdev/semp/session";

import type { Client } from "./client.js";

/** Threshold matching impl/go's `session.RekeyThreshold` constant (0.8). */
const RekeyThreshold = 0.8;

/**
 * Schedule a single rekey at 80% of the session TTL. Returns a
 * cancel function that clears the pending timer; safe to invoke any
 * number of times.
 */
export function autoRekey(client: Client): () => void {
  if (client.session === null) {
    return () => undefined;
  }
  const ttlMs = client.session.sessionTTL * 1000;
  const delay = Math.max(1000, Math.floor(ttlMs * RekeyThreshold));

  let cancelled = false;
  const handle = setTimeout(() => {
    void runRekey(client).catch((err) => {
      client.logger.warn(
        { err: err instanceof Error ? err.message : String(err) },
        "rekey failed",
      );
    });
  }, delay);

  return () => {
    if (cancelled) {
      return;
    }
    cancelled = true;
    clearTimeout(handle);
  };
}

async function runRekey(client: Client): Promise<void> {
  const session = client.session;
  if (session === null) {
    return;
  }
  if (session.isExpired()) {
    client.logger.warn("session expired before rekey");
    return;
  }
  const newId = await rekeyClient(session);
  client.logger.info(
    {
      newSessionId: newId,
      expiresAt: session.expiresAt().toISOString(),
    },
    "session rekeyed",
  );
}

File: impl/go/internal/client/rekey.go
package client

import (
	"context"
	"time"

	"semp.dev/semp-go/session"
)

// AutoRekey starts a background goroutine that rekeys the session at 80%
// of its TTL. The goroutine exits when ctx is cancelled or the rekey fails.
func (c *Client) AutoRekey(ctx context.Context) {
	if c.session == nil {
		return
	}
	go c.autoRekey(ctx)
}

func (c *Client) autoRekey(ctx context.Context) {
	threshold := time.Duration(float64(c.session.TTL) * session.RekeyThreshold)
	timer := time.NewTimer(threshold)
	defer timer.Stop()

	select {
	case <-ctx.Done():
		return
	case <-timer.C:
	}

	if !c.session.Active(time.Now()) {
		c.Log.Warn("session expired before rekey")
		return
	}

	rekeyer := &session.Rekeyer{
		Suite:   c.Suite,
		Session: c.session,
	}
	if err := rekeyer.Rekey(ctx, c.conn); err != nil {
		c.Log.Warn("rekey failed", "err", err)
		return
	}
	c.Log.Info("session rekeyed",
		"new_session_id", c.session.ID,
		"expires_at", c.session.ExpiresAt)
}
